### PR TITLE
gh-122245: move checks for writes and shadowing of __debug__ to symtable

### LINKED ISSUE #122245: Move detection of writes and shadowing of __debug__ from compiler to symtable.
There are checks in the compiler for writes and shadowing of ``__debug__``. This can be done earlier, in the symtable, where it is somewhat simpler. 

### Linked PRs
* gh-122246
* gh-122322

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #122245.

* Issue: gh-122245

### CODE REVIEW

File: Lib/test/test_syntax.py [hunk 1]
```diff
@@ -2186,6 +2261,10 @@ def f(x: *b)
       ...
    SyntaxError: yield expression cannot be used within a type alias
 
+   >>> type __debug__ = int
+   Traceback (most recent call last):
+   SyntaxError: cannot assign to __debug__
+
```
Comment (JelleZijlstra) [inline Lib/test/test_syntax.py:2267]:
```suggestion
   >>> type __debug__ = int
   Traceback (most recent call last):
   SyntaxError: cannot assign to __debug__

   >>> class A[__debug__]: pass
   Traceback (most recent call last):
   SyntaxError: cannot assign to __debug__

```
In a similar vein

File: Misc/NEWS.d/next/Core and Builtins/2024-07-24-22-39-07.gh-issue-122245.LVa9v8.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Detection of writes to ``__debug__`` is moved from the compiler's codegen
+stage to the symtable. This means that these errors now detected even in
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Core and Builtins/2024-07-24-22-39-07.gh-issue-122245.LVa9v8.rst:2]:
```suggestion
stage to the symtable. This means that these errors are now detected even in
```

File: Misc/NEWS.d/next/Core and Builtins/2024-07-24-22-39-07.gh-issue-122245.LVa9v8.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+code that is optimized away before codegen (such as assertions with the
+:option:`-O` command line option.)
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Core and Builtins/2024-07-24-22-39-07.gh-issue-122245.LVa9v8.rst:4]:
```suggestion
:option:`-O` command line option).
```

**Review by markshannon — APPROVED** [review]
Looks good.

File: Python/symtable.c [hunk 4]
```diff
@@ -2128,7 +2189,10 @@ symtable_extend_namedexpr_scope(struct symtable *st, expr_ty e)
                 VISIT_QUIT(st, 0);
             }
 
-            return symtable_add_def_helper(st, target_name, DEF_LOCAL, ste, LOCATION(e));
+            if (!symtable_add_def_helper(st, target_name, DEF_LOCAL, ste, LOCATION(e))) {
+                VISIT_QUIT(st, 0);
+            }
+            return 1;
```
Comment (iritkatriel) [inline Python/symtable.c]:
```suggestion
            return  VISIT_QUIT(st, 1);
```

File: Python/symtable.c [hunk 5]
```diff
@@ -2139,7 +2203,10 @@ symtable_extend_namedexpr_scope(struct symtable *st, expr_ty e)
                 VISIT_QUIT(st, 0);
             }
 
-            return symtable_add_def_helper(st, target_name, DEF_GLOBAL, ste, LOCATION(e));
+            if (!symtable_add_def_helper(st, target_name, DEF_GLOBAL, ste, LOCATION(e))) {
+                VISIT_QUIT(st, 0);
+            }
+            return 1;
```
Comment (iritkatriel) [inline Python/symtable.c]:
```suggestion
            return  VISIT_QUIT(st, 1);
```

File: Python/symtable.c [hunk 6]
```diff
@@ -2139,7 +2203,10 @@ symtable_extend_namedexpr_scope(struct symtable *st, expr_ty e)
                 VISIT_QUIT(st, 0);
             }
 
-            return symtable_add_def_helper(st, target_name, DEF_GLOBAL, ste, LOCATION(e));
+            if (!symtable_add_def_helper(st, target_name, DEF_GLOBAL, ste, LOCATION(e))) {
+                VISIT_QUIT(st, 0);
+            }
+            return  VISIT_QUIT(st, 1);
```
Comment (iritkatriel) [inline Python/symtable.c]:
```suggestion
            VISIT_QUIT(st, 1);
```

File: Python/symtable.c [hunk 7]
```diff
@@ -2128,7 +2189,10 @@ symtable_extend_namedexpr_scope(struct symtable *st, expr_ty e)
                 VISIT_QUIT(st, 0);
             }
 
-            return symtable_add_def_helper(st, target_name, DEF_LOCAL, ste, LOCATION(e));
+            if (!symtable_add_def_helper(st, target_name, DEF_LOCAL, ste, LOCATION(e))) {
+                VISIT_QUIT(st, 0);
+            }
+            return  VISIT_QUIT(st, 1);
```
Comment (iritkatriel) [inline Python/symtable.c]:
```suggestion
            VISIT_QUIT(st, 1);
```

**Review by JelleZijlstra — COMMENTED** [review]
I started a review but never submitted it, sorry. None of these suggested changes seem crucial, but consider adding them if you like.

File: Lib/test/test_syntax.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/test/test_syntax.py:2267]: Sorry, should have waited a bit longer. I'll add this.

**Pull request merged**